src/core/utils/slaac_address.cpp

```diff
@@ -280,29 +291,49 @@ void Slaac::GenerateIid(Ip6::NetifUnicastAddress &aAddress) const
 
     GetIidSecretKey(secretKey);
 
-    for (dadCounter = 0; dadCounter < kMaxIidCreationAttempts; dadCounter++)
+    for (uint16_t count = 0; count < kMaxIidCreationAttempts; count++, dadCounter++)
     {
         sha256.Start();
         sha256.Update(aAddress.mAddress.mFields.m8, BitVectorBytes(aAddress.mPrefixLength));
+
+        if (aNetworkId)
+        {
+            sha256.Update(aNetworkId, aNetworkIdLength);
+        }
+
         sha256.Update(netIface, sizeof(netIface));
         sha256.Update(reinterpret_cast<uint8_t *>(&dadCounter), sizeof(dadCounter));
         sha256.Update(secretKey.m8, sizeof(IidSecretKey));
         sha256.Finish(hash);
 
         aAddress.GetAddress().SetIid(&hash[0]);
 
-        // Exit and return the address if the IID is not reserved,
-        // otherwise, try again with a new dadCounter
+        // If the IID is reserved, try again with a new dadCounter
+        if (aAddress.GetAddress().IsIidReserved())
+        {
+            continue;
+        }
 
-        VerifyOrExit(aAddress.GetAddress().IsIidReserved(), OT_NOOP);
+        if (aDadCounter)
+        {
+            *aDadCounter = dadCounter;
+        }
+
+        // Exit and return the address if the IID is not reserved,
+        ExitNow(error = OT_ERROR_NONE);
     }
 
-    otLogWarnUtil("SLAAC: Failed to generate a non-reserved IID after %d attempts", dadCounter);
-    Random::Crypto::FillBuffer(hash, Ip6::Address::kInterfaceIdentifierSize);
-    aAddress.GetAddress().SetIid(&hash[0]);
+    otLogWarnUtil("SLAAC: Failed to generate a non-reserved IID after %d attempts", kMaxIidCreationAttempts);
+
+    if (aAllowLastRandomIid)
+    {
+        Random::Crypto::FillBuffer(hash, Ip6::Address::kInterfaceIdentifierSize);
```

Comment:
[RFC 7217](https://tools.ietf.org/html/rfc7217) states: "In the event that a DAD conflict cannot be solved (possibly after trying a number of different addresses), address configuration would fail." Should we really continue to support this?

Reply:
This PR simply respects previous code which already uses random in the last attempt. 
Besides, in the loop, it checks whether the IID generated is reserved, is it also counted as `duplicate address`, my previous impression is the address (not iid) conflicts with the others on the device itself or in the device's network.

Thoughts?